src/base/tests/test_util.py

```diff
@@ -389,6 +389,69 @@ def test_adapt_domain_view(self):
         self.assertIn("courriel", view_search_2.arch)
 
 
+class TestReplaceReferences(UnitTestCase):
+    def setUp(self):
+        self.cr.execute(
+            """
+            CREATE TABLE dummy_model(
+                             id serial PRIMARY KEY,
+                             res_id int,
+                             res_model varchar,
+                             extra varchar,
+                             CONSTRAINT uniq_constr UNIQUE(res_id, res_model, extra)
+                         );
+
+            INSERT INTO dummy_model(res_model, res_id, extra)
+                 VALUES -- the target is there with same res_id
+                        ('res.users', 1, 'x'),
+                        ('res.partner', 1, 'x'),
+
+                        -- two with same target and the target is there
+                        ('res.users', 2, 'x'),
+                        ('res.users', 3, 'x'),
+                        ('res.partner', 2, 'x'),
+
+                        -- two with same target and the target is not there
+                        ('res.users', 4, 'x'),
+                        ('res.users', 5, 'x'),
+
+                        -- target is there different res_id
+                        ('res.users', 6, 'x'),
+                        ('res.partner', 4, 'x')
+            """
+        )
+        self.orig_irs = list(util.INDIRECT_REFERENCES)
+        ir = util.IndirectReference("dummy_model", "res_model", "res_id")
+        util.INDIRECT_REFERENCES[:] = [ir]
+
+    def tearDown(self):
+        util.INDIRECT_REFERENCES[:] = self.orig_irs
```

Comment:
I don't like monkey patching the global list (which may be considered as an implementation detail)

You may just mock the `util.indirect_references()` function

```python
def ir_dummy(cr, bound_only=True):
    yield util.IndirectReference("dummy_model", "res_model", "res_id")

with mock.patch("odoo.upgrade.util.indirect_references.indirect_references", ir_dummy):
    util.replace_record_references_batch(...)
```